File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
using namespace std;
map<string, int> arr;

vector<string> solution(vector<string> players, vector<string> callings) {
    for(int i=0;i<players.size();i++){
        string name=players[i];
        int rank=i;
        arr[name]=rank;
    }
    
    for(int i=0;i<callings.size();i++){
        
        //이전 친구는 랭크 더하고
        //호명된 친구는 랭크 빼고
        int idx=arr[callings[i]];
        arr[players[idx-1]]++;
        arr[players[idx]]--;
        swap(players[idx], players[idx-1]);
        
        
    }
    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;
class Solution {
    public String[] solution(String[] players, String[] callings) {
        String[] answer = new String[players.length];
        HashMap<String, Integer> playerList = new HashMap<>();
        List<String> callingList = new ArrayList<>();
        for(int i=0;i<players.length;i++){
            playerList.put(players[i], i);
        }
        for(int i=0;i<callings.length;i++){
            callingList.add(callings[i]);
        }
        for(String calling: callingList){
            if(playerList.containsKey(calling)){
                int idx = playerList.get(calling);
                String prev = players[idx-1];
                String now = players[idx];
                playerList.put(prev, idx);
                playerList.put(now, idx-1);
                players[idx]=prev;
                players[idx-1]=now;
            }
        }
        return players;
    }
}